File: LeetCode/PathSum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int sum) {
        if (root == NULL) { // make sure it doesn't return true for partial paths because of nodes with single children
            return false;
        }
        if (root->left == NULL && root->right == NULL) { // leaf
            return (sum == root->val);
        }
        int sumRemnant = sum - root->val;
        return (hasPathSum(root->left, sumRemnant) || hasPathSum(root->right, sumRemnant));
    }
};

File: LeetCode/PathSum.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution
{
    public bool HasPathSum(TreeNode root, int sum)
    {
        if (root == null) // make sure it doesn't return true for partial paths because of nodes with single children
        {
            return false;
        }
        if (root.left == null && root.right == null) // leaf
        {
            return (sum == root.val);
        }
        int sumRemnant = sum - root.val;
        return (HasPathSum(root.left, sumRemnant) || HasPathSum(root.right, sumRemnant));
    }
};
